### DISCUSSION TITLE
Full Precision Option

### DISCUSSION THREAD

**Discussion by Haeretical** [BadDogSkyrim/PyNifly#187] [General]
Would it be possible to add full precision as an export option? 
I am getting seams between bodyparts, but found this can be solved by exporting as fbx/obj to Outfit Studio and setting it to full precision there. 
Using Nifly would speed up the workflow a lot as it otherwise works flawlessly. 
Since this in part seems to be based on Outfit Studio, maybe this could be exposed to Nifly during export?

**Comment by BadDogSkyrim** [comment]
Fallout 4?

**Comment by Haeretical** [comment]
Sorry, forgot to mention the export target, Fallout 4 is correct yes.

**Comment by BadDogSkyrim** [comment]
Can you give me 2 nifs which should fit seamlessly? I just did a go-round on this and what I have works with my nifs, but I'd like to check against yours.

**Comment by Haeretical** [comment]
Easiest option would be to just use Wonder Body for testing, that's the mesh where I discovered the issue first. https://www.nexusmods.com/fallout4/mods/19516
It was exported from 3dsmax with full precision from the original author iirc and there are no visible seams between body/neck/hands.
When I imported it into blender and exported it again the seams appeared. 
They can be hard to spot though. Best testing enviroment was a dark room with a window behind the char, which made the light shine through the seams. Some enbs make it a lot more visible as well, probably because of bloom.

Otherwise, if you need somebody to test the script, I'm happy to help out with that.

**Comment by podvornyakva** [comment]
@BadDogSkyrim Once again my friend? ^_^ Seams on the neck in Fallout 4... I'm so glad that i quit Fallout 4. Once i've done complete remesh of full body and head and landed them on Nexus - nobody seems was care about them. So i've deleted them. Both Nexus and hard drive. So lot efforts gone to hell...
@Haeretical Yes - there is lot issues with Fallout 4 head and body meshes. To solve them all you will need pass a long way. I can give you step by step instructions. It will be hard, and take a lot of your time. But at the end you get solid, export-ready, mesh.
So... are you ready to burn down a month of your own time?

**Comment by BadDogSkyrim** [comment]
It's really not as hard as all that.

1. Position the head and body so the neck seam is near the origin. (Vanilla does this.)
2. Make sure vert positions and bone weights match across the seam (this is normal).
3. Make sure the vertex normals match across the seam (Blender has an operator for this).

That's it for the naked body. If you have clothes that have neck seams (because the upper body shows) there are issues with the underarmor system, but that's a separate kettle of fish and better precision won't help you.

**Comment by podvornyakva** [comment]
> It's really not as hard as all that.

It is much tougher then i typed. Believe me, man. You can check it. Download every body type, install and look closer. Each one have noticeable seam. EACH ONE. Except "Demonica". But Demonica is very unique and lack of support. Each one, man...

**Comment by BadDogSkyrim** [comment]
Nope. None of mine have seams. The above instructions work.

**Comment by podvornyakva** [comment]
> Nope.

I can't believe you, man. Just can't. Cause i've wasted around half of year, messing with those meshes and textures. Major issue - meshes have a gap. Send me yours body-head set of meshes and i show you the gap on them. Or name those. You can check it by yourself. Import head and body to Bender. Join meshes and check the neck. You will see vertices mismatch. That, what dude talking about - tiny gap, affecting ENB reshader or whatever. And not only neck. Hand seams have a gap to. Those is VERY NOTICEABLE, after leaving Vault 111 on elevator. Female actor rising a hand and there is a hunge gap, if undressed.

**Comment by podvornyakva** [comment]
By the way, Bad. Can you do one more thing for Skyrim export. Skyrim pretends to have three morph files, for head meshes and faceparts - name.tri with emotions, nameChargen.tri, for face generation, and nameRaces.tri, for race morphs. Can you split, chargen and races morphs to individual files upon export to Skyrim?

**Comment by BadDogSkyrim** [comment]
Check out FurryFallout. 9 or so races * 2 sexes and not a neck seam in the bunch. (There's a texture problem on the male horses, but it's not what you're talking about.) You have to get the seam near the origin, get the weights right, and get the vertex normals right.

I know about the races .tri but haven't seen it used much. I could certainly check for the known races and split those off to their own tri. Might be most useful if it would work with custom races as well and I'm not sure how to cover that. Maybe any morph that ends with "Race", as the vanilla races do?

**Reply by podvornyakva** [reply]
Custom races do not use nameRaces.tri. Only humanoid and their "proxies" defined by RaceMenu. Most "new" races is just a copy of Nord race. Except Khadjit and Argonian. They are do not use *races.tri morphs. Probably if some one split races to individual races, and more important COMMUNITY will "eat" this, then that issue disappear. But lot got to be done for this. And new "RaceMenu" is at first place. Actually Flash interface send and resive calls through ActiveX component. But call stright to engine is more efficient. And there is a lot utilitarian mods, with non Flash GUI, that use stright calls. It is a matter of time - born of new interface... and new "RaceMenu".

**Comment by podvornyakva** [comment]
> Check out FurryFallout.

Check most common body replacers(CBBE, UNB). Not race mod. Common vanilla body replacer. And say it again.

You know, XPMSE bended shlongs used by decades, and nobody care about mismatching adult animations. Author of XPMSE do not care to. That "bend" even ported to Fallout 4. And, is suppose, with no changes, come to Starfield. Nobody care.

I've made complete remesh of CBBE, with full compatibility to any CBBE assets. No seams issues, no messed mesh, no issues at all and cleaned morphs. I've uploaded complete blender source file with textures and guide. And you know what? Nothing changes. Same people come back with the same questions of same topics - "why do i have a seam on a neck"...

About races. Open any vanilla head part in SSEEdit, except some hairs, and look at files appended to part. Every part have nameRaces.tri. Or you can check it in unpacked vanilla data. Brows, eyes, mouth, face additions and face itself have 3 morph files. I'm not talking about RaceMenu addon. I'm talking about core game files. Cause racemenu got to go to hell. New interface for Skyrim is coming... without deprecated Flash dependence. Like one, coming with open animation replacer, with hokeys extended or with community shaders. Community shaders, btw, allow to replace every shader in the game. So extra additional meshes, added by the RaceMenu(for overlays, and morphs), is not necessary anymore. Shaders have no limits for amount of textures, FBOs and VBOs usage. Even geometry shaders exist in community shaders. Parallax fur is on the way. While RaceMenu add meshes(and by that amount of DrawCalls to GPU) shader have only one drawcall and have no glitches on mipmap and LODs, as it does RaceMenu, by ovelaping multiply meshes. Shaders truly can make Skyrim prettier then Starfield.

Пока Кощей(Besethda) над златом чахнет(Starfield and upcoming TES VI), иванушки разобрали чулан на запчасти.

**Comment by podvornyakva** [comment]
Anyway. Look - there is a exporter for skyrim SE/AE hkx on nexus, integrated to game engine. At least it is something. But it is for SE 1.5.97. I probably will install SE also, and convert animations to work on. For export i've found Mike Rig on other site. It require havok, but i can deal with that on Linux.
https://www.nexusmods.com/skyrimspecialedition/mods/43108/?tab=posts&BH=0

**Comment by BadDogSkyrim** [comment]
I turned this into a discussion because it's gonna clutter up the issue tracking.

Now, why would I check out a bunch of common body replacers only to find out they screwed up the seams? The point is, even without full precision, if you follow a few rules you can get rid of the seam. As you know, since you've apparently done it with your CBBE replacer. (What mod is that, by the way?)

I know very well what the races tri file does. I considered using it for furry head variants, back when everyone was using the khajiit mesh, but opted for full head replacers instead. I'm willing to support it, as I said, just need to think through the design.

**Comment by podvornyakva** [comment]
> What mod is that, by the way?

I've deleted mod. It was named "W*ore Never Changes". I was angry, watching after year passed seens  publication, that same count of downloads and endorsments as i last visited. And i wipe mod out.

Well. Maybe you right. At least it is easier to me to make separate headmeshes and body sets for each race, then try to understand your code and make a pull request. I'm thinking about it right now, while pretending to make own race mod for Skyrim.


By the way. You don't need import meshes. Here - look at shots. Take a closer look at other mod shots and you will see seams everywhere.
[Wonder Body](https://staticdelivery.nexusmods.com/mods/1151/images/19516-19-1477926574.png)
[CBBE](https://staticdelivery.nexusmods.com/mods/1151/images/74785/74785-1694927300-1111578230.png)
[3BBB](https://staticdelivery.nexusmods.com/mods/1151/images/59503/59503-1649429836-1427750870.png)
[TWB](https://staticdelivery.nexusmods.com/mods/1151/images/36410/36410-1569270527-1261315198.jpeg)
Every body... and any texture can't solve that issue. It is mesh issue.
Here you can see even a gap. [8K Realistic Skin Textures](https://staticdelivery.nexusmods.com/mods/1151/images/28809/28809-1549270880-1695446617.jpeg)

**Reply by podvornyakva** [reply]
What i've done, to solve that issue.
1. Import 5 pats of body meshes(with morphs, btw). 7 actually. Two _facebone.nif meshes also. And also two 1st_person.
2. Snap vertices to each other on seams.
3. Mark seams.
4. Join 5 basic meshes, copy new mesh to buffer, undo join, paste new mesh.
5. Remove double vertices, from new mesh, delete custom normals, generate new one and translate data of custom normals to 9 meshes back.
6. Subdivide new mesh with multiresolution and paint details.
7. Bake new normal maps to individual images.
8. Fix morps.
9, 10 ... long way. Was complete mesh with eyes, mouth, brows, hairs, clean and smooth meshes, adequate morphs... everything in one blend file, including textures... gone.

**Reply by podvornyakva** [reply]
And know what? Now i do same things in Skyrim. If i would have a patience, i lay my hands on any staff on nexus, without any permission and excuses. Cause nobody care about you, while you follow their rules. And nobody care what you do, or what you've done. Once you brake their rules, they notice you, and what you have done, and scream at you. Try to brake you and cancel you. But i don't give a shit any more. It must be in rude way, and cruel, to anybody notice you, and what you have done. Still very angry...

**Comment by podvornyakva** [comment]
![Снимок экрана от 2023-11-22 00-48-16](https://github.com/BadDogSkyrim/PyNifly/assets/12873503/190ba3e7-d960-4bb2-abe1-5e797dd8758e)
Look. Same with meshes in Skyrim. Each male body have gap at hands. Old school... ^_^

**Reply by podvornyakva** [reply]
Increased hands mesh resolution just for mesh resolution it's own... Why, Jesus?!

**Reply by podvornyakva** [reply]
Look, Man. I've imported HIMBO meshes ad they come with HDT-SMP/CBPC collision meshes. Does it mean, that i can export them back, an they work in the game? And what i should do exactly, if so, to correctly export them?

**Reply by podvornyakva** [reply]
It is a mess... horrible. Waste of CPU/GPU and RAM usage. 3 triangles for 4 pixels at most rarely noticeable part of mesh . ^_^
![Снимок экрана от 2023-11-22 04-20-03](https://github.com/BadDogSkyrim/PyNifly/assets/12873503/2136d359-2ca5-4674-b398-c1ad51b0c920)
Hope i will deal with this faster then last time. I have experience in that, and your awesome tool. Thank you again, Man.

Important:
Got issue on FemaleHeadRace.tri import on top of mesh with base and chargen morphs. Orc head morph is disappear upon import.
Tip:
Import race.tri independently and tansfer by hands, if necessary.

**Reply by podvornyakva** [reply]
Holy shit... 648 bones. Oh my. 104 vanilla and 648 xpmsse. More than 6 times. I made realistic breast physic with 2 bones in blender. Can't imagine what are those bones for. Something defiantly goes wrong these days...

**Comment by Ghozgul** [comment]
Since the thread is already open for this subject, but where can I find the "hasFullPrecision" custom property ? I tried to export a mesh for Fallout 4 but I didn't see any option for the FullPrecision

**Comment by BadDogSkyrim** [comment]
If you *import* a mesh with full precision, select the object in Blender, and in the "Object Properties" section of the "Properties" window (orange box), scroll down to the bottom, you'll see a "Custom Properties" section. Open that and you'll see a hasFullPrecision property.

To export a new mesh, just create and set that property on the object in the same way.

**Reply by hxak016** [reply]
Adding "hasFullPrecision" int in Custom Properties fixes seam problems in my case. The blender transfer data custom normal operation is a must.

**Comment by Allnarta** [comment]
A bit of necro-posting, but I confirm that F4 body meshes all have slight seams on neck and hands. They are *really* slight though, as of, you have to use FOV 10  + TFC 1 and move camera *very* close to see them. But physically, they are there. Nothing is 100% seamless ever from what I saw.
